components/suggest/src/yelp.rs

```diff
@@ -142,69 +153,187 @@ impl<'a> SuggestDao<'a> {
+        if let Some(ref words) = post_yelp_modifier {
+            query_string = query_string[..query_string.len() - words.len()].trim();
+        }
+
+        let location = if query_string.is_empty() {
+            None
+        } else {
+            Some(query_string.to_string())
         };
 
         let (icon, icon_mimetype, score) = self.fetch_custom_details()?;
+        let subject = subject_tuple.unwrap();
         let builder = SuggestionBuilder {
-            subject: &subject,
-            subject_exact_match,
+            subject: &subject.0,
+            subject_exact_match: subject.0 == subject.1,
             pre_modifier,
             post_modifier,
+            need_location: location_sign.is_some() || location.is_some(),
             location_sign,
             location,
-            need_location,
             icon,
             icon_mimetype,
             score,
         };
         Ok(vec![builder.into()])
     }
 
+    /// Find the modifier for given query and modifier type.
+    /// Find from last word, if set FindFrom::Last to find_from.
+    /// It returns Option<String> that includes the found modifier.
+    fn find_modifier(
+        &self,
+        query: &str,
+        modifier_type: Modifier,
+        find_from: FindFrom,
+    ) -> Result<Option<String>> {
+        if query.is_empty() {
+            return Ok(None);
+        }
+
+        let words: Vec<_> = query.split_whitespace().collect();
+
+        for n in (1..=MAX_MODIFIER_WORDS_NUMBER).rev() {
+            let mut candidate_chunks: Box<dyn Iterator<Item = &[&str]>> = match find_from {
+                FindFrom::First => Box::new(words.chunks(n)),
+                _ => Box::new(words.rchunks(n)),
+            };
+            let candidate = candidate_chunks.next().unwrap_or(&[""]).join(" ");
+            if self.conn.query_row_and_then_cachable(
+                "
+                SELECT EXISTS (
+                    SELECT 1 FROM yelp_modifiers WHERE type = :type AND keyword = :word LIMIT 1
+                )
+                ",
+                named_params! {
+                    ":type": modifier_type,
+                    ":word": candidate.to_lowercase(),
+                },
+                |row| row.get::<_, bool>(0),
+                true,
+            )? {
+                return Ok(Some(candidate));
+            }
+        }
+
+        Ok(None)
+    }
+
+    /// Find the subject for given.
+    /// It returns Option<tuple> as follows:
+    /// (
+    ///   String: The keyword in DB (but the case is inherited by query).
+    ///   String: The query that was matched with the keyword.
+    /// )
+    fn find_subject(&self, query: &str) -> Result<Option<(String, String)>> {
+        if query.is_empty() {
+            return Ok(None);
+        }
+
+        if let Ok(keyword_lowercase) = self.conn.query_row_and_then_cachable(
+            "SELECT keyword
+             FROM yelp_subjects
+             WHERE :query || ' ' LIKE keyword || ' %'
```

Comment:
Can you do `:query BETWEEN keyword AND keyword || x'FFFF'`? IIRC that's more efficient than `LIKE`.

Here you're checking if `keyword` is a substring of `query`, right? And below it's the opposite?

Reply:
Yes, thanks!